### PR TITLE
Item 6965: add searching for data classes (source types)

### PR DESCRIPTION
**Pull request by labkey-susanh**

### CODE REVIEW

File: experiment/src/org/labkey/experiment/api/ExperimentServiceImpl.java [hunk 1]
```diff
@@ -745,27 +759,51 @@ public void indexDataClass(ExpDataClass dataClass)
...
+        }, Data.class, 1000);
+    }
+
+    private void indexDataClass(ExpDataClass expDataClass, SearchService.IndexTask task)
+    {
+        // Index all Data classes that have never been indexed OR where the data class has changed since last indexed
+        SQLFragment sql = new SQLFragment("SELECT * FROM ")
+                .append(getTinfoDataClass(), "dc")
+                .append(" WHERE dc.LSID = ?").add(expDataClass.getLSID())
+                .append(" AND (dc.lastIndexed IS NULL OR dc.lastIndexed < ? OR (dc.modified IS NOT NULL AND dc.lastIndexed < dc.modified))")
```
Comment (XingY) [inline experiment/src/org/labkey/experiment/api/ExperimentServiceImpl.java]: Is dc.lastIndexed < ? redundant with  dc.lastIndexed < dc.modified?

File: experiment/src/org/labkey/experiment/api/ExperimentServiceImpl.java [hunk 2]
```diff
@@ -745,27 +759,51 @@ public void indexDataClass(ExpDataClass dataClass)
...
+            {
+                ExpDataImpl impl = new ExpDataImpl(data);
+                impl.index(task);
+            }
+        }, Data.class, 1000);
+    }
+
+    private void indexDataClass(ExpDataClass expDataClass, SearchService.IndexTask task)
+    {
+        // Index all Data classes that have never been indexed OR where the data class has changed since last indexed
```
Comment (XingY) [inline experiment/src/org/labkey/experiment/api/ExperimentServiceImpl.java]: I'm not understanding the comment "all" Data classes here. It seems to me this function indexes a specific dataClass with matched lsid, not all data classes.

File: experiment/resources/schemas/exp.xml [hunk 3]
```diff
@@ -412,6 +412,11 @@
             <fkColumnName>category</fkColumnName>
         </fk>
       </column>
+        <column columnName="LastIndexed">
+            <description>Contains the time this material source was last updated in the search index</description>
```
Comment (XingY) [inline experiment/resources/schemas/exp.xml]: Data class probably shouldn't be called 'material source'

**Review by XingY — COMMENTED** [review]
Looks good, just a few question/comments.

File: experiment/src/org/labkey/experiment/api/ExperimentServiceImpl.java (hunk 1 above)
Comment (labkey-susanh) [reply experiment/src/org/labkey/experiment/api/ExperimentServiceImpl.java]: Yes, I think you're right.  Remnant of copying other indexing statement for data class objects. Redundant clause removed.

File: experiment/src/org/labkey/experiment/api/ExperimentServiceImpl.java (hunk 2 above)
Comment (labkey-susanh) [reply experiment/src/org/labkey/experiment/api/ExperimentServiceImpl.java]: Comment updated (also a copy-paste remnant, I believe).

File: experiment/resources/schemas/exp.xml (hunk 3 above)
Comment (labkey-susanh) [reply experiment/resources/schemas/exp.xml]: Yeah, probably not.

**Review by XingY — APPROVED** [review]

**Pull request merged**